File: Algorithm/assignment 5/ver2/ballpark.py
def read_data(): #데이터 입력
    f = open("ballpark.inp", "r")
    n, m = map(int, f.readline().split())
    ret = [[] for _ in range(n)]
    for i in range(n):
        ret[n-i-1] = list(map(int, f.readline().rstrip()))
    f.close()
    return n, m, ret

def print_data(cnt, ans, big): #데이터 출력
    f = open("ballpark.out", "w")
    f.write(str(big) + ' ' + str(cnt)+'\n')
    for i in range(cnt):
        f.write(str(ans[i][0]+1) + ' ' + str(ans[i][1]+1) + '\n')
    f.close()

def process(n, m, array):
    default = [[min(m-a, n-b) for a in range(m)] for b in range(n)]
    
    for i in range(n):
        for j in range(m):
            if array[i][j] == 1:
                default[i][j] = 0

    cnt = 0
    ans = []
    big = 0
    for i in range(n-2,-1,-1):
        for j in range(m-2,-1,-1):
            if default[i][j] != 0:
                default[i][j] = min(default[i][j+1], default[i+1][j], default[i+1][j+1]) + 1
            if big < default[i][j]:
                big = default[i][j]
                cnt = 1
                ans = [[j,i]]
            elif big == default[i][j]:
                cnt += 1
                ans.append([j,i])

    return cnt, sorted(ans), big
                    

    

n, m, ret = read_data()
cnt, ans, big = process(n, m, ret)
print_data(cnt, ans, big)

File: C++/ballpark.cpp
#include <iostream>
#include <fstream>
#include <string>
using namespace std;

int min(int a, int b);
int min(int a, int b, int c);

int main(){
    ifstream in("ballpark.inp");
    ofstream out("ballpark.out");

    int n, m;
    string str;
    in >> n >> m;

    getline(in, str);

    int** fArr = new int* [n];
    for(int i = 0; i < n; i++)
    {
        fArr[i] = new int[m];
        for(int j = 0; j < m; j++)
            fArr[i][j] = min(n-i, m-j);
    }

    for(int i = n-1; i >= 0; i--)
    {
        getline(in, str);
        for(int j = 0; j < m; j++)
            if(str[j] == '1')
                fArr[i][j] = 0;
    }

    int cnt = 0, big = 0;
    string ans = "";
    for(int i = n-2; i >= 0; i--)
    {
        for(int j = m-2; j >= 0; j--)
        {
            if(fArr[i][j] != 0)
                fArr[i][j] = min(fArr[i][j+1], fArr[i+1][j], fArr[i+1][j+1]) + 1;
            if(big < fArr[i][j])
            {
                big = fArr[i][j];
                cnt = 1;
            }
            else if(big == fArr[i][j])
                cnt += 1;
        }
    }

    for(int j = 0; j < m; j++)
    {
        for(int i = 0; i < n; i++)
        {
            if(fArr[i][j] == big)
                ans += to_string(j+1) + " " + to_string(i+1) + "\n";
        }
    }

    string ansStr = to_string(big) + " " + to_string(cnt) + "\n";
    out.write(ansStr.c_str(), ansStr.size());
    out.write(ans.c_str(), ans.size());

    for(int i = 0; i < n; i++)
        delete [] fArr[i];
    
    in.close();
    out.close();
    return 0;
}

int min(int a, int b)
{
    if(a < b)
        return a;
    else
        return b;
}

int min(int a, int b, int c)
{
    if(a <= b && a <= c)
        return a;
    else if(b <= a && b <= c)
        return b;
    else
        return c;
}